### PR TITLE
Hotfix/editor tools

### PR DESCRIPTION
**Pull request by jakubDoka**
Adds two tools to the editor.Very requested Copy tool for copying map terrain and blocks,"all" mode creates identical copy,"paste" mode only replaces tiles.Rectangle tool for drawing rectangle shapes,"fill" mode draws filled shapes. Icon for the Rectangle tool is missing,also translations for new tool descriptions.There is no limitation for selection size.

### CODE REVIEW

**Comment by Kieaer** [comment]
oh no I really need this

**Comment by BlueWolfGaming** [comment]
yes please.

File: core/src/mindustry/editor/MapView.java [hunk 1]
```diff
@@ -266,8 +270,20 @@ public void draw(){
...
+                    editor.tileCopy.y+editor.tileCopy.getSizeY()-1,
+                    scaling,false);
+        }
+        if(tool== EditorTool.rectangle && drawing){
+            drawRect(startx,starty,lastx,lasty,scaling,tool.mode==-1);
+        }
+
         if((!editor.drawBlock.isMultiblock() || tool == EditorTool.eraser) && tool != EditorTool.fill){
-            if(tool == EditorTool.line && drawing){
+            if(tool == EditorTool.line  && drawing){
```
Comment (Quezler) [inline core/src/mindustry/editor/MapView.java:286]:
```suggestion
            if(tool == EditorTool.line && drawing){
```

File: core/src/mindustry/editor/EditorTool.java [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
 import arc.struct.*;
 import arc.util.*;
 import mindustry.content.*;
+import mindustry.entities.type.TileEntity;
 import mindustry.game.*;
+import mindustry.maps.Map;
 import mindustry.world.*;
 import mindustry.world.blocks.*;
 
+import java.util.ArrayList;
```
Comment (Anuken) [inline core/src/mindustry/editor/EditorTool.java:16]: Read [CONTRIBUTING.MD](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md). Do not use standard Java collections.

File: core/src/mindustry/editor/EditorTool.java [hunk 3]
```diff
@@ -44,6 +51,115 @@ public void touchedLine(MapEditor editor, int x1, int y1, int x2, int y2){
             });
         }
     },
+    rectangle("fill"){
+
+        @Override
+        public void touchedLine(MapEditor editor, int x1, int y1, int x2, int y2){
+            if (x1>x2){
```
Comment (Anuken) [inline core/src/mindustry/editor/EditorTool.java:58]: This formatting is incorrect; all comparisons and assignments should have spaces around the operator. Reformat with the style file mentioned in `CONTRIBUTING.MD`.

File: core/src/mindustry/editor/TileCopy.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package mindustry.editor;
+
+import mindustry.world.Tile;
+
+import java.util.ArrayList;
+
+public class TileCopy{
+    public ArrayList<ArrayList<Tile>> data=new ArrayList<>();
```
Comment (Anuken) [inline core/src/mindustry/editor/TileCopy.java:8]: Why is this a 2D arraylist when it could be a 2D array?

Comment (jakubDoka) [reply core/src/mindustry/editor/TileCopy.java:8]: Hh sorry i absolutly forgot about that.Now am changing it and adding more new features so you can deny this pull request ill do another one later. Is it ok if i also make shortcuts for editor like selecting pencil by clicing P and so on?

**Comment by jakubDoka** [comment]
it should be changed now but am not sure newer done this before:3

**Comment by Anuken** [comment]
This PR still has formatting issues, and much of the code no longer works properly. I'll reimplement this when the editor gets its next update.

**Pull request closed without merging**